stop: Handle jobs that take a long time to stop

wmanley opened the pull request:
For nodes on a slow network connection you can get a buildup of
pending result uploads.  This can cause `/stop` to take a long time
to return and to potentially timeout after 55s with status code
202.  This new code handles that condition by retrying if that
happens.

drothlis commented:
Rebased onto master to fix conflicts & rewrote to use the new RetrySession mechanism.

drothlis committed:
stop: Handle jobs that take a long time to stop

For nodes on a slow network connection you can get a buildup of
pending result uploads.  This can cause `/stop` to take a long time
to return and to potentially timeout after 55s with status code
202.  This new code handles that condition by retrying if that
happens.

wmanley commented on stbt_rig.py:
```diff
@@ -1030,7 +1037,8 @@ def request(self, method, url, timeout=None, retry=None, **kwargs):
                 # Success or 4xx client error: don't retry:
-                if response.status_code < 500:
+                if (response.status_code < 500 and
+                        response.status_code not in retry_codes):
```
One issue is that it's not an error if we receive a 202: we don't want to sleep between requests because it's already long-poll.  I'll make this change

wmanley committed:
Handle 202 responses by retrying immediately

We return 202 "Accepted" from long-poll endpoints to indicate that the
requested operation has been started, but has not yet completed and the
request should be retried.  This is analogous to returning `EAGAIN` from
a syscall.

wmanley committed:
Refactor: Use `RetrySession` for `await_completion` too

Now `RetrySession` handles exit_status 202 to mean retry.

wmanley committed:
RetrySession: Fix behaviour if `timeout == 0`

This should time out immediately rather than time out never.

wmanley committed:
RetrySession: Use maximum per-request timeout of 60s

Requests to the portal should time-out in less time than this anyway.  The
risk of a longer timeout is that the connection gets dropped silently by a
some middlebox and we wait for ages when we're never going to get a
response.

wmanley commented:
Rebased on master to resolve `import` conflict in tests.
